docs/api_reference/src/quantities.tex

```diff
@@ -146,121 +464,1387 @@
+{
+  auto it =
+    std::ranges::find_if(args, [](auto arg) { return type_of(arg) == ^quantity_character; });
+  if (it != args.end()) return extract<quantity_character>(*it);
+  return ch;
+}
+
+template<@\defexposconceptnc{NamedQuantitySpec}@ Q>
+  requires requires { Q::dimension; }
+using @\exposidnc{to-dimension}@ = [:type_remove_const(type_of(^Q::dimension)):];
+
+struct @\exposidnc{quantity-spec-interface-base}@ {
+  friend consteval @\libconcept{QuantitySpec}@ auto operator*(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return @\exposidnc{clone-kind-of}@<lhs, rhs>(@\exposidnc{expr-multiply}@<derived_quantity_spec, struct dimensionless>(
+      @\exposidnc{remove-kind}@(lhs), @\exposidnc{remove-kind}@(rhs)));
+  }
+
+  friend consteval @\libconcept{QuantitySpec}@ auto operator/(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return @\exposidnc{clone-kind-of}@<lhs, rhs>(@\exposidnc{expr-divide}@<derived_quantity_spec, struct dimensionless>(
+      @\exposidnc{remove-kind}@(lhs), @\exposidnc{remove-kind}@(rhs)));
+  }
+
+  friend consteval bool operator==(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return ^decltype(lhs) == ^decltype(rhs);
+  }
+};
+
+struct @\exposidnc{quantity-spec-interface}@ : @\exposidnc{quantity-spec-interface-base}@ {
+  template<typename Self, @\libconcept{UnitOf}@<Self{}> U>
+  consteval @\libconcept{Reference}@ auto operator[](this Self self, U u)
+  {
+    return @\exposidnc{make-reference}@(self, u);
+  }
+
+  template<typename Self, typename FwdQ, @\libconcept{Quantity}@ Q = [:type_remove_cvref(^FwdQ):]>
+    requires @\exposconceptnc{QuantitySpecExplicitlyConvertibleTo}@<Q::quantity_spec, Self{}>
+  constexpr @\libconcept{Quantity}@ auto operator()(this Self self, FwdQ&& q)
+  {
+    return quantity{std::forward<FwdQ>(q).@\exposidnc{numerical-value}@, @\exposidnc{make-reference}@(self, Q::unit)};
+  }
+};
+
+inline constexpr struct @\libglobal{is_kind}@ {
+} @\libglobal{is_kind}@;
+
+template<auto Arg>
+concept @\defexposconceptnc{quantity-spec-1-arg}@ =
+  (type_of(^Arg) == ^quantity_character) &&
+  std::ranges::contains(enumerators_of(^quantity_character), value_of(^Arg));
+
+template<auto Lhs, auto Rhs>
+concept @\defexposconceptnc{quantity-spec-2-args}@ = (@\exposconceptnc{quantity-spec-1-arg}@<Lhs> && value_of(^Rhs) == is_kind) ||
+                               (@\exposconceptnc{quantity-spec-1-arg}@<Rhs> && value_of(^Lhs) == is_kind);
+
+template<int MaxArgs, auto... Args>
+concept @\defexposconceptnc{quantity-spec-args}@ =
+  (... && !@\libconcept{QuantitySpec}@<decltype(Args)>) &&
+  ((sizeof...(Args) == 0) ||  //
+   (sizeof...(Args) == 1 && @\exposconceptnc{quantity-spec-1-arg}@<Args...>) ||
+   (sizeof...(Args) == 2 && @\exposconceptnc{quantity-spec-2-args}@<Args...> && MaxArgs == 2));
```

Comment:
Yeah, it was very hard for me to see this as being open to program-defined extensions.
I suppose it's through [\[namespace.std\]p2](https://wg21.link/namespace.std#2).

The current implementation in the `master` branch
seems to require on the library's optional arguments
when an optional argument is specified.
That means that to use a program-defined extension with the library defaults,
you need to specify one such library default.

Reply:
Yes, the current library implementation probably needs some changes. I think it is wise to leave room for vendors and users to be able to add custom properties (e.g., non-negative) to quantity specifications. What do you think?

In the end it would be nice to have something like a dictionary in a quantity_spec of all the arguments passed so we could query if something was provided, and if so, what was its value.

Reply:
The same idiom I first suggested for the printing function optional arguments can now be used for template parameters.
```C++
mp_units::quantity_spec<x, y, {mp_units::quanity_character::vector}> // Good.
mp_units::quantity_spec<x, y, {mp_units::is_kind}> // Good.
mp_units::quantity_spec<x, y, {.is_vector{}}> // Better IMO.
mp_units::quantity_spec<x, y, {.is_kind{}}> // Better IMO.
mp_units::quantity_spec<x, y, {}, is_negative> // Program-defined extensions after the library's `struct`.

// Possible implementation:
struct quantity_spec_options {
  // The implementation uses [`std::is_within_lifetime`](https://wg21.link/meta.const.eval).
  union {
    std::true_type is_scalar = true;
    std::true_type is_vector;
    // ...
  };
  bool is_kind = false;
}
```

Reply:
That is an interesting idea. We can discuss it today.
The only downside is that the order for struct has to be strict, while the arguments are unordered on the current list.

Reply:
It's possible to implement today,
with an extra `{}`s for non-program-defined extensions,
and a `std::variant`-like implementation (<https://cpp1.godbolt.org/z/evW7sre6e>):
```C++
#include <variant>
#include <type_traits>


struct qty_opts {
  union character {
    std::true_type scalar;
    std::true_type vector;
    std::true_type complex;
    std::true_type tensor;
  };
  union kind {
    std::true_type is_kind;
  };
  template<class T = std::monostate>
  struct type {
    int index = 0;
    union {
      T program_defined_extension;
      character char_v;
      kind kind_v;
    };
    consteval type(T v) : program_defined_extension{v} { }
    consteval type(character c) : char_v{c} { }
    consteval type(kind k) : kind_v{k} { }
  };
};

template<qty_opts::type... O>
struct quantity_spec { };

struct is_negative { };

int main() {
  (void)quantity_spec<>{};
  // (void)quantity_spec<{}>{};
  (void)quantity_spec<{{.vector{}}}>{};
  // (void)quantity_spec<{{.complex{}, .is_kind{}}}>{};
  (void)quantity_spec<{{.is_kind{}}}>{};
  (void)quantity_spec<{is_negative{}}, {{.is_kind{}}}, {{.tensor{}}}>{};
}

```

Maybe we can take advantage of arrays
or nested anonymous unions
to avoid the extra `{}`s.
Here's using nested anonymous unions,
but that removes the `index` which makes it usable today (<https://cpp1.godbolt.org/z/8KG34r6s7>):
```C++
#include <variant>
#include <type_traits>

template<class T = std::monostate>
union qty_opts {
  T program_defined_extension;
  union {
    std::true_type scalar;
    std::true_type vector;
    std::true_type complex;
    std::true_type tensor;
  };
  std::true_type is_kind;
};

template<qty_opts... O>
struct quantity_spec { };

struct is_negative { };

int main() {
  (void)quantity_spec<>{};
  (void)quantity_spec<{}>{};
  (void)quantity_spec<{.vector{}}>{};
  // (void)quantity_spec<{.complex{}, .is_kind{}}>{};
  (void)quantity_spec<{.is_kind{}}>{};
  (void)quantity_spec<{is_negative{}}, {.is_kind{}}, {.tensor{}}>{};
}
```